Add vision pipeline diagram to documentation

cat101 opened the pull request:
Understanding the relationship between all the concepts was something that prevented me from understanding the configuration file and the meaning of some key settings. I drew this diagram to help me understand. 

I'm not sure if it is complete and whether you think it would fit on the configuration section of the docs. If you give me some guidance/direction I can add to it

Commit:
Add vision pipeline diagram

NickM-27 commented:
I think the Frigate introduction or some other intro page would be a better location for this, the config reference is already very long and this is not directly related (just helpful).

Could also host this on a separate page and then link to it from multiple places

NickM-27 commented:
Overall though this is really cool and will be helpful for people to understand

Commit:
Moved pipeline to the Guides section

cat101 commented:
Thanks @NickM-27 for the feedback. I have moved the diagrams to the "Guides" section. I have also added a bit more detail to the pipelines and therefore I have also prepared a high level diagram. Can you please check the detailed diagram? I want to make sure it is correct before I add some text to explain it

Thanks

NickM-27 commented:
Thanks, a few things:

1. I would recommend removing go2rtc. go2rtc doesn't have anything to do with the video pipeline directly (as far as frigate is concerned, it doesn't care where the stream comes from when ffmepg decodes it)
2. I think the "motion clips" area should be reworked. motion segments are only kept when continuous recording is set to `motion` mode. or when `all` mode is configured, every segment is kept. So for one I think motion clips should likely be called motion segments, and perhaps add some text to explain that only applies when the recording retention is configured as motion
3. birdseye does not use motion / object detection snapshots; it pulls directly from all of the decoded frames but only shows when motion / objects are detected (again depending on configuration). I think you could simply change "motion snapshot" and "detection snapshot" could be changed to "___ detected" but there may be a better way to align with how my point `2.` is shown.
4. Instead of clip store I would suggest calling it recording store, clips are not really stored directly
5. I would add another item between "object detect" and "track objects" which is `Apply object filters`

Commit:
Fixing segment storage and other flows

cat101 commented:
Thanks @NickM-27 for the feedback and clarifications. I have accommodated all of them except for the first one which I want to dig deeper. As a newcomer to Frigate I did not know of go2rtc, so when I first read the documentation I did not know why I would need it... It turned out that my cheap camera did not give a clean RTSP source so I spent a lot of time modding its firmware. Now I know that this camera uses the dvr163 protocol and I could have just used go2rtc as a frontend to Frigate.....  

The way I see tfrigate & go2rtc moving forward is that Frigate will focus being the best NVR while go2rtc will focus on being able to interface with more & more cameras. This is why I want to bring go2rtc to the "foreground" .... BTW: I understand the go2rtc has a role in streaming, but on these area Frigate has more options (jsmpeg, birdseye, etc)...   

What do you think? How do we present go2rtc's role for a newcomer?

NickM-27 commented:
I think with it being optional it's fine. Overall I think this looks good

Commit:
Added text

NickM-27 commented on docs/docs/guides/video_pipeline.md:
```diff
@@ -2,26 +2,29 @@
+This guide provides an overview to help users understand some of the key Frigate concepts.
 
-### High level view of the video pipeline
+### Overview
```
```suggestion
## Overview
```

cat101 replied:
While this is conceptually right, I think it makes the navigation harded/weirder on the right bar. If we had more sections and sub sections I would agree, but I think that for this simple doc I would leave both sections as "equals"

NickM-27 replied:
I am suggesting that all sections should be two, maybe I missed some

NickM-27 replied:
the problem is that when future edits are made to the docs, the number of `#` is just incremented and then it leads to the section not having a side section at all.

cat101 replied:
Ahhh I missed that... done

NickM-27 commented on docs/docs/guides/video_pipeline.md:
```diff
@@ -2,26 +2,29 @@
+As the diagram shows, all feeds first need to be acquired. Depending on the data source, it may be as simple as using FFmpeg to connect to an RTSP source via TCP or something more involved like connecting to an Apple Homekit camera using go2rtc. A single camera can produce a main (i.e. high quality) and a sub (i.e. low quality) video feed. Typically, the sub-feed will be decoded to produce full-f ...
 ### Detailed view of the video pipeline
 
+The following diagram adds a lot more detail than the simple view explained before. The goal is to show the detailed data paths between the processing steps.
```
```suggestion
As the diagram shows, all feeds first need to be acquired. Depending on the data source, it may be as simple as using FFmpeg to connect to an RTSP source via TCP or something more involved like connecting to an Apple Homekit camera using go2rtc. A single camera can produce a main (i.e. high resolution) and a sub (i.e. lower resolution) video feed. 

Typically, the sub-feed will be decoded to produce full-frame images. As part of this process, the resolution may be downscaled and an image sampling frequency may be imposed (e.g. keep 5 frames per second). 

These frames will then be compared over time to detect movement areas (a.k.a. motion boxes). These motion boxes are combined into motion regions and are analyzed by a machine learning model to detect known objects. Finally, the snapshot and recording retention config will decide what video clips and events should be saved.

## Detailed view of the video pipeline

The following diagram adds a lot more detail than the simple view explained before. The goal is to show the detailed data paths between the processing steps.
```

cat101 replied:
LGTM

cat101 commented:
@NickM-27 I have worked the go2rtc part to make it more clear why it is there. I have also added text to explain the diagram I think it is now ready for publishing. Please give it a quick check and I'll merge it.

Thanks for all your help refining the ideas!

cat101 committed:
Update docs/docs/guides/video_pipeline.md

Co-authored-by: Nicolas Mowen <nickmowen213@gmail.com>

Commit:
Keeping case as other pages

Commit:
Changing sections to ##

cat101 commented:
I realized I cannot merge. Should I close the PR or will you take it from here?

NickM-27 commented:
Blake will review as well and merge once it looks good

blakeblackshear commented on docs/docs/guides/video_pipeline.md:
```diff
@@ -0,0 +1,65 @@
+%%{init: {"themeVariables": {"edgeLabelBackground": "transparent"}}}%%
+
+flowchart LR
+    Feed(Feed\nAcquisition) --> Decode(Video\ndecoding)
+    Decode --> Motion(Motion\nDetection)
+    Motion --> Object(Object\nDetection)
+    Feed --> Recording(Recording\n&\nVisualization)
+    Motion --> Recording
+    Object --> Recording
```
```suggestion
    Feed(Feed\nacquisition) --> Decode(Video\ndecoding)
    Decode --> Motion(Motion\ndetection)
    Motion --> Object(Object\ndetection)
    Feed --> Recording(Recording\nand\nvisualization)
    Motion --> Recording
    Object --> Recording
```

blakeblackshear commented on docs/docs/guides/video_pipeline.md:
````diff
@@ -0,0 +1,65 @@
+```mermaid
+%%{init: {"themeVariables": {"edgeLabelBackground": "transparent"}}}%%
+
+flowchart TD
+    RecStore[(Recording\nstore)]
+    SnapStore[(Snapshot\nstore)]
+
+    subgraph Aquisition
+        Cam["Camera"] -->|FFmpeg supported| Stream
+        Cam -->|"Other streaming\nprotocols"| go2rtc
+        go2rtc("go2rtc") --> Stream
+        Stream[Capture main & Sub\nstreams] --> |detect stream|Decode(Decode & Downscale)
+    end
+    subgraph Motion
+        Decode --> MotionM(Apply\nmotion masks)
+        MotionM --> MotionD(Motion\ndetection)
+    end
+    subgraph Detection
+        MotionD --> |motion regions| ObjectD(Object detection)
+        Decode --> ObjectD
+        ObjectD --> ObjectFilter(Apply object filters & zones)
+        ObjectFilter --> ObjectZ(Track objects)
+    end
+    Stream --> |decoded frames|BirdsEye
+    MotionD --> |motion event|BirdsEye
+    ObjectZ --> |object event|BirdsEye
+
+    MotionD --> |"video segments\n(retain motion)"|RecStore
+    ObjectZ --> |detection clip|RecStore
+    Stream -->|"video segments\n(retain all)"| RecStore
+    ObjectZ --> |detection snapshot|SnapStore
````
```suggestion
flowchart TD
    RecStore[(Recording\nstore)]
    SnapStore[(Snapshot\nstore)]

    subgraph Acquisition
        Cam["Camera"] -->|FFmpeg supported| Stream
        Cam -->|"Other streaming\nprotocols"| go2rtc
        go2rtc("go2rtc") --> Stream
        Stream[Capture main and\nsub streams] --> |detect stream|Decode(Decode and\ndownscale)
    end
    subgraph Motion
        Decode --> MotionM(Apply\nmotion masks)
        MotionM --> MotionD(Motion\ndetection)
    end
    subgraph Detection
        MotionD --> |motion regions| ObjectD(Object detection)
        Decode --> ObjectD
        ObjectD --> ObjectFilter(Apply object filters & zones)
        ObjectFilter --> ObjectZ(Track objects)
    end
    Decode --> |decoded frames|Birdseye
    MotionD --> |motion event|Birdseye
    ObjectZ --> |object event|Birdseye

    MotionD --> |"video segments\n(retain motion)"|RecStore
    ObjectZ --> |detection clip|RecStore
    Stream -->|"video segments\n(retain all)"| RecStore
    ObjectZ --> |detection snapshot|SnapStore
```

cat101 committed:
Update docs/docs/guides/video_pipeline.md

Co-authored-by: Blake Blackshear <blakeb@blakeshome.com>

cat101 committed:
Update docs/docs/guides/video_pipeline.md

Co-authored-by: Blake Blackshear <blakeb@blakeshome.com>

cat101 commented:
Thanks @blakeblackshear . The changes look good. I have merged them
